File: langs/javascript/src/interpolation.ts
/**
 * Variable interpolation engine for DOTLYTE.
 *
 * Supports:
 *   ${VAR}           — substitute with value of VAR
 *   ${VAR:-default}  — fallback if VAR is unset or empty
 *   ${VAR:?error}    — throw if VAR is unset
 *   $$               — literal $ escape
 *
 * Resolution order: same file (top-to-bottom) → already-merged config → env vars.
 * Circular references are detected and throw InterpolationError.
 */

import { InterpolationError } from "./errors.js";

/** Regex matching ${VAR}, ${VAR:-default}, ${VAR:?message} */
const INTERPOLATION_RE = /\$\{([^}]+)\}/g;

/**
 * Interpolate all `${VAR}` references in a flat key-value map.
 *
 * @param data     — raw key-value pairs from dotenv parsing (pre-coercion)
 * @param context  — already-resolved config values from higher-priority sources
 * @param envVars  — process.env for fallback resolution
 * @returns A new map with all interpolations resolved.
 */
export function interpolate(
  data: Record<string, string>,
  context: Record<string, unknown> = {},
  envVars: Record<string, string | undefined> = process.env,
): Record<string, string> {
  const result: Record<string, string> = {};
  const resolving = new Set<string>();

  function resolve(key: string): string {
    // Already resolved in this pass
    if (key in result) return result[key]!;

    // Circular reference detection
    if (resolving.has(key)) {
      throw new InterpolationError(
        `Circular variable reference detected: ${[...resolving, key].join(" → ")}`,
        key,
      );
    }

    const raw = data[key];
    if (raw === undefined) {
      // Not in this file — check context > env
      const fromContext = flatGet(context, key);
      if (fromContext !== undefined) return String(fromContext);
      const fromEnv = envVars[key] ?? envVars[key.toUpperCase()];
      if (fromEnv !== undefined) return fromEnv;
      return "";
    }

    resolving.add(key);
    result[key] = expandValue(raw, resolve, context, envVars);
    resolving.delete(key);

    return result[key]!;
  }

  // Resolve all keys (order = insertion order = file order)
  for (const key of Object.keys(data)) {
    if (!(key in result)) {
      resolve(key);
    }
  }

  return result;
}

/**
 * Expand a single value string, replacing all ${...} references.
 */
function expandValue(
  value: string,
  resolve: (key: string) => string,
  context: Record<string, unknown>,
  envVars: Record<string, string | undefined>,
): string {
  // Handle literal $$ → $
  let escaped = value.replace(/\$\$/g, "\x00DOLLAR\x00");

  escaped = escaped.replace(INTERPOLATION_RE, (_match, expr: string) => {
    // ${VAR:?error message}
    if (expr.includes(":?")) {
      const [varName, ...errParts] = expr.split(":?");
      const errMsg = errParts.join(":?").trim();
      const resolved = resolveVar(varName!.trim(), resolve, context, envVars);
      if (resolved === undefined || resolved === "") {
        throw new InterpolationError(
          errMsg || `Required variable '${varName!.trim()}' is not set`,
          varName!.trim(),
        );
      }
      return resolved;
    }

    // ${VAR:-default}
    if (expr.includes(":-")) {
      const [varName, ...defaultParts] = expr.split(":-");
      const defaultVal = defaultParts.join(":-");
      const resolved = resolveVar(varName!.trim(), resolve, context, envVars);
      return resolved !== undefined && resolved !== "" ? resolved : defaultVal;
    }

    // ${VAR}
    const resolved = resolveVar(expr.trim(), resolve, context, envVars);
    return resolved ?? "";
  });

  return escaped.replace(/\x00DOLLAR\x00/g, "$");
}

/**
 * Resolve a variable name from: same-file data → context → env vars.
 */
function resolveVar(
  name: string,
  resolve: (key: string) => string,
  context: Record<string, unknown>,
  envVars: Record<string, string | undefined>,
): string | undefined {
  // Try same-file first (will recurse into resolve())
  try {
    const fromFile = resolve(name);
    if (fromFile !== "") return fromFile;
  } catch {
    // Not in file — fall through
  }

  // Try context (already-merged config)
  const fromContext = flatGet(context, name);
  if (fromContext !== undefined) return String(fromContext);

  // Try env vars (case-insensitive fallback)
  const fromEnv = envVars[name] ?? envVars[name.toUpperCase()];
  if (fromEnv !== undefined) return fromEnv;

  return undefined;
}

/**
 * Dot-notation or flat lookup in a nested object.
 */
function flatGet(obj: Record<string, unknown>, key: string): unknown {
  // Direct key first
  if (key in obj) return obj[key];

  // Dot-notation traversal
  const parts = key.split(".");
  let current: unknown = obj;
  for (const part of parts) {
    if (current !== null && typeof current === "object" && !Array.isArray(current)) {
      current = (current as Record<string, unknown>)[part];
    } else {
      return undefined;
    }
  }
  return current;
}

File: langs/python/src/dotlyte/interpolation.py
"""Variable interpolation engine for DOTLYTE.

Expands ${VAR}, ${VAR:-default}, ${VAR:?error} references in .env values.
Detects circular references.
"""

from __future__ import annotations

import re
from typing import Any

from dotlyte.errors import InterpolationError

# Pattern: ${VAR}, ${VAR:-default}, ${VAR:?error}
_VAR_PATTERN = re.compile(r"\$\{([^}]+)\}")


def interpolate(
    data: dict[str, str],
    context: dict[str, Any] | None = None,
    env_vars: dict[str, str] | None = None,
) -> dict[str, str]:
    """Expand variable references in a dictionary of string values.

    Resolution order: same-file → context → os.environ.

    Args:
        data: Raw string key-value pairs from .env file.
        context: Already-loaded config values for resolution.
        env_vars: Optional env vars dict (defaults to os.environ).

    Returns:
        New dict with all variable references expanded.

    Raises:
        InterpolationError: On circular references or undefined required vars.

    """
    import os

    ctx = context or {}
    envs = env_vars if env_vars is not None else dict(os.environ)
    resolved: dict[str, str] = {}
    resolving: set[str] = set()

    def _resolve_key(key: str) -> str:
        if key in resolved:
            return resolved[key]
        if key in resolving:
            chain = " → ".join(resolving) + f" → {key}"
            raise InterpolationError(
                f"Circular variable reference detected: {chain}",
                key=key,
            )

        if key not in data:
            # Fall back to context or env
            ctx_val = _get_nested(ctx, key)
            if ctx_val is not None:
                return str(ctx_val)
            env_key = key.upper()
            if env_key in envs:
                return envs[env_key]
            return ""

        resolving.add(key)
        val = _expand(data[key])
        resolving.discard(key)
        resolved[key] = val
        return val

    def _expand(value: str) -> str:
        # Handle $$ escape
        value = value.replace("$$", "\x00DOLLAR\x00")

        def _replacer(match: re.Match[str]) -> str:
            expr = match.group(1)

            # ${VAR:?error}
            if ":?" in expr:
                var_name, error_msg = expr.split(":?", 1)
                val = _lookup(var_name.strip())
                if not val:
                    raise InterpolationError(
                        error_msg.strip() or f"Required variable '{var_name.strip()}' is not set",
                        key=var_name.strip(),
                    )
                return val

            # ${VAR:-default}
            if ":-" in expr:
                var_name, default = expr.split(":-", 1)
                val = _lookup(var_name.strip())
                return val if val else default

            # ${VAR}
            return _lookup(expr.strip())

        result = _VAR_PATTERN.sub(_replacer, value)
        return result.replace("\x00DOLLAR\x00", "$")

    def _lookup(var_name: str) -> str:
        # Check same-file data first
        lower = var_name.lower()
        if lower in data:
            return _resolve_key(lower)
        # Check context
        ctx_val = _get_nested(ctx, lower)
        if ctx_val is not None:
            return str(ctx_val)
        # Check env vars
        upper = var_name.upper()
        if upper in envs:
            return envs[upper]
        return ""

    for key in data:
        if key not in resolved:
            _resolve_key(key)

    return resolved


def _get_nested(data: dict[str, Any], key: str) -> Any:
    """Get a value from a nested dict using dot-notation or flat key."""
    # Try flat key first
    if key in data:
        return data[key]
    # Try dot-notation
    parts = key.split(".")
    current: Any = data
    for part in parts:
        if isinstance(current, dict) and part in current:
            current = current[part]
        else:
            return None
    return current
